File: server/routes/alerts.js
import express from 'express'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { dirname } from 'path'
import { v4 as uuidv4 } from 'uuid'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const ALERTS_PATH = path.join(__dirname, '../data/alerts.json')
const SETTINGS_PATH = path.join(__dirname, '../data/alert-settings.json')

const router = express.Router()

const readAlerts = () => JSON.parse(fs.readFileSync(ALERTS_PATH, 'utf-8'))
const writeAlerts = (data) => fs.writeFileSync(ALERTS_PATH, JSON.stringify(data, null, 2), 'utf-8')
const readSettings = () => JSON.parse(fs.readFileSync(SETTINGS_PATH, 'utf-8'))
const writeSettings = (data) => fs.writeFileSync(SETTINGS_PATH, JSON.stringify(data, null, 2), 'utf-8')

// 심각도 계산
const getSeverity = (totalScore, threshold) => {
  if (totalScore < threshold - 10) return 'danger'   // 위험 🔴
  if (totalScore < threshold) return 'warning'        // 주의 🟡
  return null
}

// 알림 생성 (history.js에서 호출)
export const createAlertIfNeeded = (record) => {
  try {
    const settings = readSettings()
    const alerts = readAlerts()
    const { totalScore, violence, language, sexual, title, channelTitle, thumbnail, videoId, profileId, watchedAt } = record

    const reasons = []
    const threshold = settings.threshold

    if (totalScore < threshold) reasons.push(`종합 점수 ${totalScore}점`)
    if (violence !== undefined && violence < threshold) reasons.push(`폭력성 ${violence}점`)
    if (language !== undefined && language < threshold) reasons.push(`언어 ${language}점`)
    if (sexual !== undefined && sexual < threshold) reasons.push(`선정성 ${sexual}점`)

    // 늦은 시간 감지
    const hour = new Date(watchedAt).getHours()
    const isLateNight = settings.lateNightAlert && hour >= settings.lateNightHour
    if (isLateNight) reasons.push(`늦은 시간(${hour}시) 시청`)

    if (reasons.length === 0) return null

    const severity = isLateNight && totalScore >= threshold ? 'warning' : getSeverity(totalScore, threshold)

    // 같은 영상 + 프로필 알림이 이미 있으면 반복 시청으로 업데이트
    const existing = alerts.find(a => a.videoId === videoId && a.profileId === profileId)
    if (existing) {
      existing.watchCount = (existing.watchCount || 1) + 1
      existing.repeated = true
      existing.updatedAt = watchedAt
      writeAlerts(alerts)
      return existing
    }

    const newAlert = {
      id: uuidv4(),
      profileId,
      videoId,
      title,
      channelTitle,
      thumbnail,
      totalScore,
      violence,
      language,
      sexual,
      reasons,
      severity: severity || 'warning',
      watchedAt,
      watchCount: 1,
      repeated: false,
      read: false,
    }

    alerts.unshift(newAlert)
    writeAlerts(alerts)
    return newAlert
  } catch (err) {
    console.error('알림 생성 실패:', err)
    return null
  }
}

// 알림 전체 조회
router.get('/', (req, res) => {
  try {
    const alerts = readAlerts()
    res.json({ alerts })
  } catch {
    res.status(500).json({ error: '알림을 불러오는 중 오류가 발생했어요' })
  }
})

// 알림 읽음 처리 (단건)
router.patch('/:id/read', (req, res) => {
  try {
    const alerts = readAlerts()
    const alert = alerts.find(a => a.id === req.params.id)
    if (!alert) return res.status(404).json({ error: '알림을 찾을 수 없어요' })
    alert.read = true
    writeAlerts(alerts)
    res.json({ success: true })
  } catch {
    res.status(500).json({ error: '읽음 처리 중 오류가 발생했어요' })
  }
})

// 알림 전체 읽음 처리
router.patch('/read-all', (req, res) => {
  try {
    const alerts = readAlerts()
    alerts.forEach(a => { a.read = true })
    writeAlerts(alerts)
    res.json({ success: true })
  } catch {
    res.status(500).json({ error: '전체 읽음 처리 중 오류가 발생했어요' })
  }
})

// 알림 설정 조회
router.get('/settings', (req, res) => {
  try {
    res.json(readSettings())
  } catch {
    res.status(500).json({ error: '설정을 불러오는 중 오류가 발생했어요' })
  }
})

// 알림 설정 저장
router.put('/settings', (req, res) => {
  try {
    const { threshold, lateNightAlert, lateNightHour } = req.body
    const current = readSettings()
    const updated = {
      threshold: threshold ?? current.threshold,
      lateNightAlert: lateNightAlert ?? current.lateNightAlert,
      lateNightHour: lateNightHour ?? current.lateNightHour,
    }
    writeSettings(updated)
    res.json({ success: true, settings: updated })
  } catch {
    res.status(500).json({ error: '설정 저장 중 오류가 발생했어요' })
  }
})

export default router

File: server/routers/alerts.py
"""
위험 영상 알림 라우터 — Supabase DB 버전 (JSON 파일에서 이전)

멀티테넌시: user_id 스코프.
- create_alert_if_needed: history 저장 시 호출 (async, user_id 필요)
- alert_settings 는 user 당 1행 (upsert)
DB 컬럼 snake_case → 프론트 camelCase 변환 (_to_api).
"""

from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional

from auth import get_current_user
from db import sb_select, sb_insert, sb_update, sb_upsert

router = APIRouter()

# 알림 설정 기본값 (user 설정이 없을 때)
DEFAULT_SETTINGS = {"threshold": 70, "lateNightAlert": True, "lateNightHour": 22}


def _to_api(row: dict) -> dict:
    return {
        "id": row.get("id"),
        "profileId": row.get("profile_id"),
        "videoId": row.get("video_id"),
        "title": row.get("title"),
        "channelTitle": row.get("channel_title"),
        "thumbnail": row.get("thumbnail"),
        "totalScore": row.get("total_score"),
        "violence": row.get("violence"),
        "language": row.get("language"),
        "sexual": row.get("sexual"),
        "reasons": row.get("reasons") or [],
        "severity": row.get("severity"),
        "watchedAt": row.get("watched_at"),
        "watchCount": row.get("watch_count"),
        "repeated": row.get("repeated"),
        "read": row.get("read"),
        "updatedAt": row.get("updated_at"),
    }


def get_severity(total_score: int, threshold: int) -> Optional[str]:
    if total_score is None:
        return None
    if total_score < threshold - 10:
        return "danger"
    if total_score < threshold:
        return "warning"
    return None


async def _read_settings(user_id: str) -> dict:
    """user 의 알림 설정을 camelCase 로 반환 (없으면 기본값)."""
    rows = await sb_select(
        "alert_settings",
        {"user_id": f"eq.{user_id}", "select": "*", "limit": "1"},
    )
    if not rows:
        return dict(DEFAULT_SETTINGS)
    r = rows[0]
    return {
        "threshold": r.get("threshold", 70),
        "lateNightAlert": r.get("late_night_alert", True),
        "lateNightHour": r.get("late_night_hour", 22),
    }


async def create_alert_if_needed(record: dict, user_id: str) -> Optional[dict]:
    """history 저장 시 호출 — 위험/늦은시간 조건이면 알림 생성·갱신."""
    try:
        settings = await _read_settings(user_id)

        total_score = record.get("totalScore")
        violence = record.get("violence")
        language = record.get("language")
        sexual = record.get("sexual")
        video_id = record.get("videoId")
        profile_id = record.get("profileId")
        watched_at = record.get("watchedAt")

        threshold = settings.get("threshold", 70)
        reasons = []

        if total_score is not None and total_score < threshold:
            reasons.append(f"종합 점수 {total_score}점")
        if violence is not None and violence < threshold:
            reasons.append(f"폭력성 {violence}점")
        if language is not None and language < threshold:
            reasons.append(f"언어 {language}점")
        if sexual is not None and sexual < threshold:
            reasons.append(f"선정성 {sexual}점")

        # 늦은 시간 감지
        hour = datetime.fromisoformat(watched_at.replace("Z", "+00:00")).hour if watched_at else 0
        late_night_alert = settings.get("lateNightAlert", False)
        late_night_hour = settings.get("lateNightHour", 22)
        is_late_night = late_night_alert and hour >= late_night_hour
        if is_late_night:
            reasons.append(f"늦은 시간({hour}시) 시청")

        if not reasons:
            return None

        severity = "warning" if (is_late_night and (total_score or 0) >= threshold) else (get_severity(total_score, threshold) or "warning")

        # 같은 영상 + 프로필 알림이 이미 있으면 반복 시청으로 업데이트
        existing = await sb_select(
            "alerts",
            {
                "user_id": f"eq.{user_id}",
                "video_id": f"eq.{video_id}",
                "profile_id": f"eq.{profile_id}",
                "select": "id,watch_count",
                "limit": "1",
            },
        )
        if existing:
            row = existing[0]
            updated = await sb_update(
                "alerts",
                {"id": f"eq.{row['id']}"},
                {
                    "watch_count": (row.get("watch_count") or 1) + 1,
                    "repeated": True,
                    "updated_at": watched_at,
                },
            )
            return _to_api(updated[0]) if updated else None

        inserted = await sb_insert("alerts", {
            "user_id": user_id,
            "profile_id": profile_id,
            "video_id": video_id,
            "title": record.get("title"),
            "channel_title": record.get("channelTitle"),
            "thumbnail": record.get("thumbnail"),
            "total_score": total_score,
            "violence": violence,
            "language": language,
            "sexual": sexual,
            "reasons": reasons,
            "severity": severity,
            "watched_at": watched_at,
            "watch_count": 1,
            "repeated": False,
            "read": False,
        })
        return _to_api(inserted[0]) if inserted else None

    except Exception as e:
        print(f"알림 생성 실패: {e}")
        return None


# GET /alerts
@router.get("")
async def get_alerts(user: dict = Depends(get_current_user)):
    rows = await sb_select(
        "alerts",
        {"user_id": f"eq.{user['user_id']}", "select": "*", "order": "created_at.desc"},
    )
    return {"alerts": [_to_api(a) for a in rows]}


# PATCH /alerts/read-all
@router.patch("/read-all")
async def read_all_alerts(user: dict = Depends(get_current_user)):
    await sb_update("alerts", {"user_id": f"eq.{user['user_id']}"}, {"read": True})
    return {"success": True}


# PATCH /alerts/{id}/read
@router.patch("/{alert_id}/read")
async def read_alert(alert_id: str, user: dict = Depends(get_current_user)):
    updated = await sb_update(
        "alerts",
        {"id": f"eq.{alert_id}", "user_id": f"eq.{user['user_id']}"},
        {"read": True},
    )
    if not updated:
        raise HTTPException(status_code=404, detail="알림을 찾을 수 없어요")
    return {"success": True}


# GET /alerts/settings
@router.get("/settings")
async def get_settings(user: dict = Depends(get_current_user)):
    return await _read_settings(user["user_id"])


class SettingsUpdate(BaseModel):
    threshold: Optional[int] = None
    lateNightAlert: Optional[bool] = None
    lateNightHour: Optional[int] = None


# PUT /alerts/settings
@router.put("/settings")
async def update_settings(data: SettingsUpdate, user: dict = Depends(get_current_user)):
    current = await _read_settings(user["user_id"])
    merged = {
        "threshold": data.threshold if data.threshold is not None else current["threshold"],
        "lateNightAlert": data.lateNightAlert if data.lateNightAlert is not None else current["lateNightAlert"],
        "lateNightHour": data.lateNightHour if data.lateNightHour is not None else current["lateNightHour"],
    }
    await sb_upsert("alert_settings", {
        "user_id": user["user_id"],
        "threshold": merged["threshold"],
        "late_night_alert": merged["lateNightAlert"],
        "late_night_hour": merged["lateNightHour"],
    }, on_conflict="user_id")
    return {"success": True, "settings": merged}
